### ISSUE TITLE
`nextest` dependency could be handled with other prerequisite dependencies

### ISSUE THREAD

**Issue by jordanhendricks** [oxidecomputer/omicron#3944]
I noticed while looking at #3939 that `cargo nextest` is handled differently from other dependency requirements for developing omicron: Instead of it being automatically downloaded as a part of the normal "prerequisite" script flow, the repo-level README mentions downloading or installing it specifically before running the prerequisite scripts. For CI, the binary is downloaded explicitly, similar to other dependencies:
https://github.com/oxidecomputer/omicron/blob/3412cb17e5d52796ad5a81020992d6a851f89f92/.github/buildomat/build-and-test.sh

I could understand wanting to give developers some flexibility here in terms of how this package is installed. At the least, I think our tooling should be checking that `cargo nextest` is on the system, perhaps in terms of PATH, like we have done for other expected binaries:
https://github.com/oxidecomputer/omicron/blob/3412cb17e5d52796ad5a81020992d6a851f89f92/tools/install_builder_prerequisites.sh#L204-L211

I think ideally the prerequisite tools (such as the one I am working on for #3939) would also install this dependency, as that would remove a step required to do a pretty common development task (running tests). I am curious if anyone disagrees on that point.

**Comment by davepacheco** [comment]
Thanks for raising this.  I think it's important that we have a golden path that reliably works (and is tested in CI) and that it includes `nextest`.  I can also understand if folks want to manage these dependencies separately.  Someone could always choose to skip this step (and take responsibility for setting up the dependencies themselves).  Depending on how much work we want to put into supporting this, I could imagine having the tool accept arguments saying what to set up, like:

```
cargo xtask prereqs all                  # installs build-time and run-time dependencies
cargo xtask prereqs build                # installs all build-time dependencies
cargo xtask prereqs build --no-nextest # installs build-time dependencies except for nextest
cargo xtask prereqs nextest              # installs only nextest
cargo xtask prereqs cockroach            # installs only cockroach
...
```

We could have whatever aliases make sense there.  I could imagine something like `internal` that installs the Oxide-developed stuff (console, maghemite, dendrite, transceiver_control) but not the third party stuff.

I'm not sure if it's worth putting in the work to do all this, but if it's a Rust program, it might be pretty easy to do.

This approach probably makes it more important to validate (like it sounds like we already do) that the tools we need are on your PATH and match the version we expect before we go to use them.

What do you think?

**Comment by sunshowers** [comment]
I really like the overall suggestions here, and agree that doing them as part of a Rust program (as opposed to bash) will make it easier to implement conditional logic.

**Comment by sunshowers** [comment]
The reason I didn't go down this path was essentially that there are lots of ways to install nextest (e.g. directly, via homebrew, via cargo-binstall etc), and I figured it was quite complex to nail down all the different paths across the different OSes and so on. I think at least a check to ensure nextest is available and up-to-date (`cargo nextest show-config version`) would be useful.

**Comment by jordanhendricks** [comment]
@davepacheco definitely! Installing specific dependencies is have something I have built into the structure of #3939 already. The current interface looks slightly different than that, with:
- `cargo xtasks prereqs pkg <system package name>` for system packages installed with the OS/distro level package manager
- `cargo xtasks prereqs dep <dep name, such as nextest or cockroach>` for deps that get installed outside of the package manager. Currently I believe that is mostly artifacts from buildomat or downloads from a known source URL, and I would slot nextest in there.

We could instead make the `cargo xtasks install <pkg / dep>` a flat namespace, which might be more convenient. Any strong opinions there?

@sunshowers I agree that it is hard today to add a new dep like this, particularly one that doesn't look similar to the other dependencies installed by `ci_download_<dep name>`, such as cargo-nextest, and I also agree that as you pointed out, this is an interesting case in the prereqs dependencies because there are several ways to install it, some of which might be preferred by different folks. (I could imagine for example, that you as the maintainer of nextest, would appreciate being able to point the omicron repository at a development copy of nextest.)

A check for cargo-nextests existence and version as you suggest can certainly be incorporated into the `cargo xtasks prereqs check` subcommand. I would also like to have a "golden path" that will install it as a part of `cargo xtasks prereqs install alll` subcommand, if possible. In the current implementation of the prereqs xtask, this command will install all packages and deps for your detected OS, though the dependencies part is not yet implemented, and I could imagine building in some flexibility there to tweak how one installs dependencies, which could allow overriding the choice the tool makes for how to install cargo-nextest, for example. Another idea floated by @davepacheco to me offline was an option to the `prereqs` tools that allows one to *exclude* a dep from the "all deps" list, which could be useful for a situation like this.

I can scope out how complex these options look for #3939 and come up with a strawman implementation there to start with!
